// nodejs-polars/bin/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.INSPECT_SYMBOL = exports.regexToString = exports.isIterator = exports.isExprArray = exports.isSeriesArray = exports.isDataFrameArray = exports.range = exports.isPath = exports.selectionToExprList = exports.columnOrColumnsStrict = exports.columnOrColumns = void 0;
const expr_1 = require("./lazy/expr");
const series_1 = require("./series/series");
const dataframe_1 = require("./dataframe");
const path_1 = __importDefault(require("path"));
const types_1 = require("util/types");
function columnOrColumns(columns) {
    if (columns) {
        return columnOrColumnsStrict(columns);
    }
}
exports.columnOrColumns = columnOrColumns;
function columnOrColumnsStrict(...columns) {
    return columns.flat(3);
}
exports.columnOrColumnsStrict = columnOrColumnsStrict;
function selectionToExprList(columns, stringToLit) {
    return [columns].flat(3).map(expr => (0, expr_1.exprToLitOrExpr)(expr, stringToLit)._expr);
}
exports.selectionToExprList = selectionToExprList;
function isPath(s, expectedExtensions) {
    const { base, ext, name } = path_1.default.parse(s);
    return Boolean(base && ext && name) && !!(expectedExtensions?.includes(ext));
}
exports.isPath = isPath;
const range = (start, end) => {
    const length = end - start;
    return Array.from({ length }, (_, i) => start + i);
};
exports.range = range;
const isDataFrameArray = (ty) => Array.isArray(ty) && dataframe_1.DataFrame.isDataFrame(ty[0]);
exports.isDataFrameArray = isDataFrameArray;
const isSeriesArray = (ty) => Array.isArray(ty) && series_1.Series.isSeries(ty[0]);
exports.isSeriesArray = isSeriesArray;
const isExprArray = (ty) => Array.isArray(ty) && expr_1.Expr.isExpr(ty[0]);
exports.isExprArray = isExprArray;
const isIterator = (ty) => ty !== null && typeof ty[Symbol.iterator] === "function";
exports.isIterator = isIterator;
const regexToString = (r) => {
    if ((0, types_1.isRegExp)(r)) {
        return r.source;
    }
    return r;
};
exports.regexToString = regexToString;
exports.INSPECT_SYMBOL = Symbol.for("nodejs.util.inspect.custom");
